### ISSUE TITLE
Firebase Storage - Files with custom metadata keys containing spaces and colons can longer be accessed

### ISSUE THREAD

**Issue by NroL** [firebase/firebase-unity-sdk#208]
**Operating system:** Android 8+ & iOS 11+ (productionized mobile app), Unity + Windows (development)
**Operating system version:** Unity 2019.3.15f1, Windows 10

**What steps will reproduce the problem:**
1. Create a basic Unity file with Firebase SDK v8.4.0 (just the required Storage + other mandatory Firebase components)
2. Upload a few images using Firebase Storage containing custom metadata _keys_ with spaces and colons next to each other (eg. `CustomKey :  `). Put whatever for the _values_.
4. Go to Firebase web console and view the uploaded images in the Firebase Storage web dashboard.

**What is the expected result?**
Inside the Firebase Storage web dashboard, we should be able to :
1. click on the uploaded image to view the image preview on the right-hand panel 
2. click on the hyperlinked filename ("Name") to automatically open a new webpage that shows the full image

Separately, the uploaded content can be downloaded back into our mobile app via the default Firebase Storage instructions.

**What happens instead of that?**
Inside the Firebase Storage web dashboard, I instead get:
1. "Error loading preview" when an image is selected, instead of an image preview
2. clicking on the hyperlinked filename ("Name") automatically opens up a new webpage that states the following error:

> "502. That’s an error.
> The server encountered a temporary error and could not complete your request.
> Please try again in 30 seconds. That’s all we know."

When attempting to download the storage content in Unity, I see this error in Unity logs:
> "Failed to download because System.AggregateException: One or more errors occurred. ---> Firebase.Storage.StorageException: The server did not return a valid JSON response.  Contact Firebase support if this issue persists.
>    --- End of inner exception stack trace ---
> ---> (Inner Exception #0) Firebase.Storage.StorageException: The server did not return a valid JSON response.  Contact Firebase support if this issue persists.<---"

**Please provide any other information below, and attach a screenshot if possible.**

We have not changed our Firebase code in the last 2-3 months, and we just started seeing this problem recently. 
This issue affects all uploaded content with custom metadata keys containing spaces and colons - all the way back from 2020 when we started using Firebase Storage.

**Additional info:**
- While the image in Firebase Storage dashboard is having issues, the same uploaded image can still be accessed via Google Cloud Console's storage dashboard without any issues (if using _Authenticated URL)_.

**Comment by NroL** [comment]
Update - additional info:

I mentioned previously that the images cause Error 502 when accessed from Firebase Storage dashboard, and that the same images can be accessed without issues through the Google Cloud Platform's storage dashboard.

I just found out that the images can be accessed via the GCP _only_ through their _Authenticated URL_ (default).
When I changed an image's access to public, the new _Public URL_ gives the same Error 502, similar to when accessed via Firebase Storage dashboard.

**Comment by NroL** [comment]
Quick update - I have identified the cause of the issue (relating to custom metadata of files + permissions) and currently applying a fix.

Will post a more detailed followup and close this off after the fix is successfully applied.

**Comment by NroL** [comment]
**Final update**

Our issues were caused by the combination of the following: 
1. some of our files' custom metadata keys contain a space and a colon next to each other.
2. using firebase storage for these files 

As an example, the custom metadata key `CustomKey : ` will not work as it has both a space and a colon next to each other, while `CustomKey:` will still work. 

A few more notes on these files:
- They **can** still be accessed/downloaded via Google Cloud Storage if we use an _Authenticated URL_. 
- They **cannot** be accessed/downloaded via Google Cloud Storage if we use the _Public URL_.
- They **cannot** be accessed/downloaded via Firebase Storage,.

As a solution, I used a python script to crawl through all our files in Google Cloud Storage and modify the problematic custom metadata keys to remove all spaces and colons.

I initially tried to use gsutil's _setmeta_ command ([link](https://cloud.google.com/storage/docs/gsutil/commands/setmeta)) to remove the custom metadata keys with `:` because it would have been much quicker. 
However, the setmeta command already uses `:` as a way to split metadata key and value, and I couldn't find a way to put a literal `:` inside the key itself.
